### PR TITLE
Improved tooltip system unregistering to fix reparenting hack

### PR DESCRIPTION
**Pull request by hhyyrylainen**
**Brief Description of What This PR Does**
I reworked the logic for the tooltips a bit to move more to a manual unregistering model to make things work when some objects need to support scene detach and re-attach later.

**Related Issues**

This is a follow up to: https://github.com/Revolutionary-Games/Thrive/pull/4284

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Review by Kasterisk — CHANGES_REQUESTED** [review]
Logic seems fine but I commented on some things.

File: src/engine/NodeHelpers.cs
Comment (hhyyrylainen) [reply src/engine/NodeHelpers.cs:9]: I'd rather just remove all traces of this meta workaround. If there is a big need to return it, it is in git history. And anyway this is simple enough to just quickly wing it again if needed.

File: src/engine/PostShutdownActions.cs
Comment (hhyyrylainen) [reply src/engine/PostShutdownActions.cs:4]: I'll add that.

File: src/gui_common/art_gallery/GalleryViewer.cs
Comment (hhyyrylainen) [reply src/gui_common/art_gallery/GalleryViewer.cs:455]: Yeah, but the overall system everywhere is a bit confused with the difference between what is a tooltip and what's the tooltip data. I think *I* kind of understand this now, but in the code "tooltip" is used to refer mostly to the tooltip data that is registered for controls. I'll rename this "UnregisterToolTips".

File: src/gui_common/charts/line/LineChart.cs
Comment (hhyyrylainen) [reply src/gui_common/charts/line/LineChart.cs:274]: This method is named like this because it is actually not capable of creating new tooltips, it can only reapply existing registrations that are on hold.

File: src/gui_common/charts/line/LineChart.cs
Comment (hhyyrylainen) [reply src/gui_common/charts/line/LineChart.cs:1108]: I'll try to come up with something else... like I said it's already confusing that both tooltips and the tooltip data / callback registrations are called the same thing. A new refactor to solve that would be needed, and I think the simpler term "tooltip" should be reserved for other code, the tooltip manager itself should use a more complicated term to mean the tooltip nodes.

File: src/gui_common/tooltip/ToolTipCallbackData.cs
Comment (hhyyrylainen) [reply src/gui_common/tooltip/ToolTipCallbackData.cs:47]: Well not really, because for reparenting reasons a thing shouldn't remove the tooltip registrations unless it is not going to be used anymore. So for example a custom control that releases its tooltips when it is deleted, would incorrectly lead to warnings printed here if that node was reparented. That's why I added the PostShutdownActions to detect the condition of forgotten tooltips.

Comment (Kasterisk) [reply src/gui_common/tooltip/ToolTipCallbackData.cs:47]: > That's why I added the PostShutdownActions to detect the condition of forgotten tooltips.

I guess that's good enough.

File: src/gui_common/charts/line/LineChart.cs
Comment (Kasterisk) [reply src/gui_common/charts/line/LineChart.cs]: I admit that I terribly named these things, e.g the name tooltip data might be confused as the content in that tooltip, for which callback handlers would more accurate.

> the tooltip manager itself should use a more complicated term to mean the tooltip nodes.

I don't entirely agree with that... If only there's a more shorter and concise form to describe a temporary object which contain GUI callback functions for the tooltipable controls.

I don't have anything in mind so let's leave it.

**Comment by hhyyrylainen** [comment]
I didn't get that to happen. It shouldn't really happen. What did you do to cause that to happen?

**Comment by Kasterisk** [comment]
Just standard testing procedure. Started microbe stage, enter the editor, return, repeat. I also went to freebuild mode in the same session and finally closed the game with the window's close button (not quit button).

**Comment by hhyyrylainen** [comment]
I'll re-test this. I just played the game in freebuild for a couple of editor cycles before exiting the game, and I didn't hit the breakpoint I had put on the warning print about unreleased tooltips.

**Review by Kasterisk — APPROVED** [review]
In any case, this looks good to me now.

**Comment by hhyyrylainen** [comment]
I tried those steps you described again but I again did not hit the `Some tooltips have not been unregistered on game shutdown` error. I'll merge this once the last check passes.

**Pull request merged**
